src/k8s-engine/__tests__/k3sHelper.spec.ts

```diff
@@ -265,4 +266,129 @@ describe(K3sHelper, () => {
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+      test('picks same minor version before pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3', 'v1.2.9', 'v1.3.0']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('picks same minor version after pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3', 'v1.5.0', 'v1.5.9']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+      test('picks same patch version before pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s8', 'v1.2.4+k3s1']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('picks same patch version after pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.4+k3s1', 'v1.2.4+k3s9']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+
+      test('picks same premajor version before pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s2', 'v1.2.4+k3s1']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('can choose closest pre-release version for single digits', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s3', 'v1.2.3+k3s4']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+      test('verify single digits prereleases can pick the lower one', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s2', 'v1.2.3+k3s3', 'v1.2.3+k3s5']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('can choose closest pre-release version when crossing from single digits into 10s', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s3', 'v1.2.3+k3s8', 'v1.2.3+k3s12']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+    });
+
+    describe('selectClosestSemVer', () => {
+      // Note filenames are specified deliberately out of order in order to test sorting
+      test('finds the closest major version', () => {
+        const desiredSemver = new semver.SemVer('v3.1.2+k3s3');
+        const cachedFilenames = ['v1.2.9+k3s4', 'v4.2.8+k3s1', 'v4.3.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[1]);
+      });
+      test('finds the closest minor version', () => {
+        const desiredSemver = new semver.SemVer('v1.12.2+k3s3');
+        const cachedFilenames = ['v1.23.9+k3s4', 'v2.12.8+k3s1', 'v1.7.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[2]);
```

Comment:
We wanted `1.12`, but got `1.7` instead? Wouldn't it be better to not do downgrades (since that involves wiping data)?
Of course, finding a matching version is even better…

Reply:
If we want 1.12, and have to choose between 1.7 and 1.23, I picked 1.7 because it's closer numerically.

Reply:
I agree with @mook-as that selecting the closest higher version sounds safer.

It is a lot more likely that you rely on a newer feature, that may be missing in the older version, than that you would be broken by an incompatible change in a later one.